nbconvert/exporters/qtpdf.py

```diff
@@ -0,0 +1,120 @@
+"""Export to PDF via a headless browser"""
+
+# Copyright (c) IPython Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import asyncio
+
+import tempfile, os
+
+from traitlets import Bool, default
+from jupyter_core.paths import jupyter_path
+import concurrent.futures
+
+from .html import HTMLExporter
+
+
+class QtPDFExporter(HTMLExporter):
+    """Writer designed to write to PDF files.
+
+    This inherits from :class:`HTMLExporter`. It creates the HTML using the
+    template machinery, and then uses pyqtwebengine to create a pdf.
+    """
+    export_from_notebook = "PDF via HTML"
+
+    paginate = Bool(
+        True,
+        help="""
+        Split generated notebook into multiple pages.
+
+        If False, a PDF with one long page will be generated.
+
+        Set to True to match behavior of LaTeX based PDF generator
+        """
+    ).tag(config=True)
+
+    output_mimetype = "application/pdf"
+
+    @default('file_extension')
+    def _file_extension_default(self):
+        return '.pdf'
+
+    @default('template_name')
+    def _template_name_default(self):
+        return 'qtpdf'
+
+    @default('template_data_paths')
+    def _template_data_paths_default(self):
+        return jupyter_path("nbconvert", "templates", "qtpdf")
+
+    def _check_launch_reqs(self):
+        try:
+            from PyQt5 import QtCore, QtWidgets, QtWebEngineWidgets
+            from PyQt5.QtGui import QPageLayout, QPageSize
+        except ModuleNotFoundError as e:
+            raise RuntimeError("PyQtWebEngine is not installed to support Qt PDF conversion. "
+                               "Please install `nbconvert[qtpdf]` to enable.") from e
+        return QtCore, QtWidgets, QtWebEngineWidgets, QPageLayout, QPageSize
+
+    def run_pyqtwebengine(self, html):
+        """Run pyqtwebengine."""
+
+        ext = ".html"
+        temp_file = tempfile.NamedTemporaryFile(suffix=ext, delete=False)
+        pdf_filename = f"{temp_file.name[:-len(ext)]}.pdf"
+        with temp_file:
+            temp_file.write(html.encode('utf-8'))
+
+        try:
+            QtCore, QtWidgets, QtWebEngineWidgets, QPageLayout, QPageSize = self._check_launch_reqs()
+            app = QtWidgets.QApplication([""])
+            loader = QtWebEngineWidgets.QWebEngineView()
+            loader.page().pdfPrintingFinished.connect(
+                lambda *args: app.exit())
+            loader.load(QtCore.QUrl(f"file://{temp_file.name}"))
+
+            def emit_pdf(dimensions):
+                if self.paginate:
+                    page_size = QPageSize(QPageSize.A4)
+                    page_layout = QPageLayout(page_size, QPageLayout.Portrait, QtCore.QMarginsF())
+                else:
+                    page_size = QPageSize(QtCore.QSizeF(*dimensions), QPageSize.Point)  # FIXME
```

Comment:
I couldn't get it to work, the returned dimensions from the JavaScript seem wrong. Also, probably needs a conversion from pixels to points.